source/carton-bindings-ocaml/src/lib.rs

```diff
@@ -0,0 +1,54 @@
+mod export;
+mod tensor;
+
+use crate::export::Tensor;
+
+use carton_core::error::CartonError;
+use carton_core::types::{Device, LoadOpts, Tensor as CartonTensor};
+use carton_core::Carton;
+use ocaml_rust::{Custom, CustomConst};
+use tokio::runtime::Runtime;
+
+/// Load a carton model
+fn load(
+    path: String,
+    visible_device: Option<String>,
+    override_runner_name: Option<String>,
+    override_required_framework_version: Option<String>,
+) -> Result<CustomConst<Carton>, Custom<CartonError>> {
+    Runtime::new().unwrap().block_on(async {
```

Comment:
Is there a way for us to expose async functionality to OCaml from Rust?

I've been exploring the same thing for C and C++ and am writing a doc that should be in an upcoming PR with either the C or C++ bindings.

Reply:
I am no Ocaml god, and am mainly just interested in learning Ocaml. Most of the async in ocaml that I have seen is by using a specific library and preprocessor (`Lwt`) as Ocaml does not nativity understand what async is. There would have to be some way to return a Ocaml type but also call functions to the async library from rust. I don't know if that is possible, and if it is I would not even know how to start doing that.

Once you get the C/C++ bindings I will check how you explore async, and will see if I can do something similar.

Reply:
Is [this](https://opensource.janestreet.com/async/) something that's widely used? I haven't written OCaml in a very long time so I don't have much context.

Here are the C bindings: #169 (the PR is larger than I hoped; sorry :/)

The doc I mentioned is not included in that PR although the included README.md provides an overview of the async approach for C.

My C++ PR will talk about design tradeoffs a little, but the short version is that I'm likely going to have async Rust functions return `std::future`s in C++ and also provide something like `CartonAsyncNotifier` from the C interface (because polling several `std::future`s is inefficient).

If there isn't native async/coroutine/green threads/event loop functionality in OCaml, maybe we can do something similar to what I'm planning to do for C++?

Reply:
I have never interacted with the janestreet async library, so I am guessing its not *super* widely used. It looks very similar to Lwt though. With Ocaml I have not done much of writing async functions and have only really used them, so I will have to do some research on how you can make your own async functions with Lwt to see if it is possible to add support through ffi.

As for the initial of your c ffi, it looks like you are making your own version of async. I will have to see if I can implement the Ocaml async with a bigger library as having the "normal" feel to it would be a lot better. If it is not possible then I will probably just have to do what you did.

Reply:
>  as having the "normal" feel to it would be a lot better.

Yes, that's generally a priority for the bindings. It should feel like it "belongs" in the language.

The callback + opaque user-provided argument approach is fairly common in C, but we shouldn't directly translate that to other languages unless there isn't a better alternative.

Reply:
For what it's worth, I decided not to use `cxx` for the C++ bindings.

It doesn't natively support some features that are important for making the C++ API feel normal (namely template instantiation). For some other parts of the API, it would either require extra copies or require us to write wrapper C++ code on top of the generated `cxx` code to avoid these copies.

So if we need to write wrapper C++ code and/or create (fairly tedious) workarounds to get a normal feeling API, we might as well implement idiomatic C++ bindings on top of the C API.

`cxx` will probably be useful for creating bindings for the metadata structs (e.g. [`CartonInfo`](https://github.com/VivekPanyam/carton/blob/f208de72c85512c600a8f84997be9c9a4c74f18e/source/carton/src/info.rs#L50)), but doesn't seem like it's worth it for the rest of the API.

Reply:
The C++ PR is up at #174. It's also unfortunately quite large, but the included `README.md` might be worth taking a look at